File: 0086-分隔链表/Solution.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

class Solution {
    public ListNode partition(ListNode head, int x) {
        if(head == null || head.next == null){
            return head;
        }
        ListNode nhead = new ListNode(Integer.MIN_VALUE);
        nhead.next = head;
        ListNode small = nhead;
        ListNode big = nhead;
        while(small.next != null){
            if(small.next.val < x){
                small = small.next;
            }else{
                big = small.next;
                break;
            }
        }
        if(big == nhead){
            return head;
        }
        ListNode work = big.next;
        while(work != null){
            if(work.val >= x){
                big = work;
                work = work.next;
            }else{
                big.next = work.next;
                work.next = small.next;
                small.next = work;
                small = small.next;
                work = big.next;
            }
        }
        return nhead.next;
    }
}

File: 0086-分隔链表/Solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        if not head:
            return head
        list1, list2 = [], []
        work = head
        while work:
            if work.val < x:
                list1.append(work.val)
            else:
                list2.append(work.val)
            work = work.next
        list1.extend(list2)
        work = head
        for val in list1:
            work.val = val
            work = work.next
        return head
        